src/volume_calc.cpp

```diff
@@ -90,182 +123,236 @@ std::vector<VolumeCalculation::Result> VolumeCalculation::execute() const
+          * (volume_sample - result.volume[0]) / total_samples);
+        result.iterations = iterations;
+
+        // update threshold value if needed
+        if (trigger_type_ != VolumeTriggerMetric::NONE) {
+          double val = 0.0;
+          switch (trigger_type_) {
+            case VolumeTriggerMetric::STD_DEV:
+              val = result.volume[1];
+              break;
+            case VolumeTriggerMetric::REL_ERR:
+              val = result.volume[0] == 0.0 ? 0.0 : result.volume[1] / result.volume[0];
+              break;
+            case VolumeTriggerMetric::VARIANCE:
+              val = result.volume[1] * result.volume[1];
               break;
-            }
+          }
+          // update max if entry is valid
+          if (val > 0.0) { trigger_val = std::max(trigger_val, val); }
+        }
+
+        for (int j = 0; j < n_nuc; ++j) {
+          // Determine total number of atoms. At this point, we have values in
+          // atoms/b-cm. To get to atoms we multiply by 10^24 V.
+          double mean = 1.0e24 * volume_sample * atoms(j, 0);
+          double stdev = 1.0e24 * volume_sample * std::sqrt(atoms(j, 1));
+
+          // Convert full arrays to vectors
+          if (mean > 0.0) {
+            result.nuclides.push_back(j);
+            result.atoms.push_back(mean);
+            result.uncertainty.push_back(stdev);
           }
         }
       }
-    } else {
-      int q = master_indices[i_domain].size();
-      int buffer[2*q];
-      for (int k = 0; k < q; ++k) {
-        buffer[2*k] = master_indices[i_domain][k];
-        buffer[2*k + 1] = master_hits[i_domain][k];
-      }
+    } // end domain loop
 
-      MPI_Send(&q, 1, MPI_INTEGER, 0, 0, mpi::intracomm);
-      MPI_Send(&buffer[0], 2*q, MPI_INTEGER, 0, 1, mpi::intracomm);
-    }
-#endif
+    // if no trigger is applied, we're done
+    if (trigger_type_ == VolumeTriggerMetric::NONE) { return results; }
 
+#ifdef OPENMC_MPI
+    // update maximum error value on all processes
     if (mpi::master) {
-      int total_hits = 0;
-      for (int j = 0; j < master_indices[i_domain].size(); ++j) {
-        total_hits += master_hits[i_domain][j];
-        double f = static_cast<double>(master_hits[i_domain][j]) / n_samples_;
-        double var_f = f*(1.0 - f)/n_samples_;
-
-        int i_material = master_indices[i_domain][j];
-        if (i_material == MATERIAL_VOID) continue;
-
-        const auto& mat = model::materials[i_material];
-        for (int k = 0; k < mat->nuclide_.size(); ++k) {
-          // Accumulate nuclide density
-          int i_nuclide = mat->nuclide_[k];
-          atoms(i_nuclide, 0) += mat->atom_density_[k] * f;
-          atoms(i_nuclide, 1) += std::pow(mat->atom_density_[k], 2) * var_f;
-        }
+      for (int i = 1; i < mpi::n_procs; i++) {
+        MPI_Send(&trigger_val, 1, MPI_DOUBLE, i, 0, mpi::intracomm);
       }
+    } else {
+      MPI_Recv(&trigger_val, 1, MPI_DOUBLE, 0, 0, mpi::intracomm, MPI_STATUS_IGNORE);
```

Comment:
This block here can just be an `MPI_Bcast`

Reply:
`MPI_Bcast` is a collective operation that needs to be called by all processes. You can remove the `MPI_Recv` as well as the if-block altogether, and just have the single line calling `MPI_Bcast`

Reply:
Hasty changes.. thanks!